**MAESTRÍA EN INTELIGENCIA ARTIFICIAL APLICADA**

**Curso: TC5053 - Ciencia y analítica de datos**

Tecnológico de Monterrey

Prof Grettel Barceló Alonso

**Semana 2**
Pandas para el análisis de datos en Python

---

*   NOMBRE: Carlos Humberto Galván Perales
*   MATRÍCULA: A01797969


---

En esta actividad trabajarás con el archivo `cleaned_weather.csv`, un extracto del conjunto de datos meteorológicos a lo largo de todo el año 2020 en una estación del Instituto *Max Planck* (Alemania) disponible en Kaggle.

Los datos meteorológicos fueron registrados cada 10 minutos e incluyen los siguientes indicadores:

*   `timestamp`: Fecha y hora de la observación.
*   `p`: Presión atmosférica en milibares (mbar)
*   `T`: Temperatura del aire en grados Celsius (°C)
*   `Tpot`: Temperatura potencial en Kelvin (K)
*   `rh`: Humedad relativa en porcentaje (%)
*   `VPact`: Presión real de vapor en milibares (mbar)
*   `sh`: Humedad específica en gramos por kilogramo (g/kg)
*   `H2OC`: Concentración de vapor de agua en milimoles por mol (mmol/mol) de aire seco
*   `rho`: Densidad del aire en gramos por metro cúbico (g/m³)
*   `wv`: Velocidad del viento en metros por segundo (m/s)
*   `wd`: Dirección del viento en grados (°)
*   `rain`: Precipitación total en milímetros (mm)
*   `raining`: Duración de la lluvia en segundos (s)

**NOTA IMPORTANTE:** Asegúrate de responder *explícitamente* todos los cuestionamientos.


In [1]:
# Importa las librerías necesarias
import pandas as pd

1.	Descarga el archivo: `cleaned_weather.csv` y guarda, en un dataframe (`weather_df`), todos sus registros.
- Determina la dimensionalidad del dataframe y obtén los identificadores de columnas.
- Renombra las columnas para facilitar la interpretación de los indicadores en los ejercicios siguientes:
  - `medicion, presion_atmosferica, temperatura_celsius, temperatura_celsius, temperatura_kelvin, humedad_relativa, presion_vapor, humedad_especifica, concentracion_vapor, densidad_aire, velocidad_viento, direccion_viento, precipitacion_total, duracion_lluvia`
- Muestra los primeros y los últimos 5 registros.
- ¿Hay valores faltantes en el dataframe?





In [ ]:
# Cargar el archivo cleaned_weather.csv en un DataFrame llamado weather_df
weather_df = pd.read_csv("cleaned_weather.csv")

# Determinar la dimensionalidad del DataFrame
print(f"Dimensionalidad del DataFrame, Filas: {weather_df.shape[0]} Columnas: {weather_df.shape[1]}")

# Mostrar los nombres de las columnas de forma legible
print("Identificadores de columnas:")
for col in weather_df.columns:
    print(f" - {col}")

# Renombrar las columnas para facilitar la interpretación
weather_df.columns = [
    'medicion',                # timestamp
    'presion_atmosferica',     # p
    'temperatura_celsius',     # T
    'temperatura_kelvin',      # Tpot
    'humedad_relativa',        # rh
    'presion_vapor',           # VPact
    'humedad_especifica',      # sh
    'concentracion_vapor',     # H2OC
    'densidad_aire',           # rho
    'velocidad_viento',        # wv
    'direccion_viento',        # wd
    'precipitacion_total',     # rain
    'duracion_lluvia'          # raining
]

# Mostrar los primeros 5 registros
print("Primeros 5 registros:")
print(weather_df.head())

# Mostrar los últimos 5 registros
print("Últimos 5 registros:")
print(weather_df.tail())

# Verificar si hay valores faltantes en el DataFrame
print("¿Hay valores faltantes en el DataFrame?")
print(weather_df.isnull().sum())


In [2]:
# Load DataFrame
weather_df = pd.read_csv("cleaned_weather.csv")

In [3]:
# Verify DataFrame shape
weather_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52696 entries, 0 to 52695
Data columns (total 13 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   timestamp  52696 non-null  object 
 1   p          52696 non-null  float64
 2   T          52696 non-null  float64
 3   Tpot       52696 non-null  float64
 4   rh         52696 non-null  float64
 5   VPact      52696 non-null  float64
 6   sh         52696 non-null  float64
 7   H2OC       52696 non-null  float64
 8   rho        52696 non-null  float64
 9   wv         52696 non-null  float64
 10  wd         52696 non-null  float64
 11  rain       52696 non-null  float64
 12  raining    52696 non-null  int64  
dtypes: float64(11), int64(1), object(1)
memory usage: 5.2+ MB


In [4]:
# Column names
weather_df.columns

Index(['timestamp', 'p', 'T', 'Tpot', 'rh', 'VPact', 'sh', 'H2OC', 'rho', 'wv',
       'wd', 'rain', 'raining'],
      dtype='object')

In [5]:
# Renombrar las columnas para facilitar la interpretación
weather_df.columns = [
    'medicion',                # timestamp
    'presion_atmosferica',     # p
    'temperatura_celsius',     # T
    'temperatura_kelvin',      # Tpot
    'humedad_relativa',        # rh
    'presion_vapor',           # VPact
    'humedad_especifica',      # sh
    'concentracion_vapor',     # H2OC
    'densidad_aire',           # rho
    'velocidad_viento',        # wv
    'direccion_viento',        # wd
    'precipitacion_total',     # rain
    'duracion_lluvia'          # raining
]



**Mostrar los primeros y ultimos 5 registros**

Esto permite darnos una idea general de los valores y columnas en el DataFrame.

In [6]:
# Mostrar los primeros 5 registros
weather_df.head()

medicion  presion_atmosferica  temperatura_celsius  \
0  01/01/20 0:10              1008.89                 0.71   
1  01/01/20 0:20              1008.76                 0.75   
2  01/01/20 0:30              1008.66                 0.73   
3  01/01/20 0:40              1008.64                 0.37   
4  01/01/20 0:50              1008.61                 0.33   

   temperatura_kelvin  humedad_relativa  presion_vapor  humedad_especifica  \
0              273.18              86.1           5.54                3.42   
1              273.22              85.2           5.49                3.39   
2              273.21              85.1           5.48                3.39   
3              272.86              86.3           5.41                3.35   
4              272.82              87.4           5.47                3.38   

   concentracion_vapor  densidad_aire  velocidad_viento  direccion_viento  \
0                 5.49        1280.62              1.02             224.3   
1                 5.45        1280.33              0.43             206.8   
2                 5.43        1280.29              0.61             197.1   
3                 5.37        1281.97              1.11             206.4   
4                 5.42        1282.08              0.49             209.6   

   precipitacion_total  duracion_lluvia  
0                  0.0                0  
1                  0.0                0  
2                  0.0                0  
3                  0.0                0  
4                  0.0                0

In [7]:
# Mostrar los primeros 5 registros, pero transponiendo para poder observar todas las columnas del dataframe
weather_df.head().T

0              1              2  \
medicion             01/01/20 0:10  01/01/20 0:20  01/01/20 0:30   
presion_atmosferica        1008.89        1008.76        1008.66   
temperatura_celsius           0.71           0.75           0.73   
temperatura_kelvin          273.18         273.22         273.21   
humedad_relativa              86.1           85.2           85.1   
presion_vapor                 5.54           5.49           5.48   
humedad_especifica            3.42           3.39           3.39   
concentracion_vapor           5.49           5.45           5.43   
densidad_aire              1280.62        1280.33        1280.29   
velocidad_viento              1.02           0.43           0.61   
direccion_viento             224.3          206.8          197.1   
precipitacion_total            0.0            0.0            0.0   
duracion_lluvia                  0              0              0   

                                 3              4  
medicion             01/01/20 0:40  01/01/20 0:50  
presion_atmosferica        1008.64        1008.61  
temperatura_celsius           0.37           0.33  
temperatura_kelvin          272.86         272.82  
humedad_relativa              86.3           87.4  
presion_vapor                 5.41           5.47  
humedad_especifica            3.35           3.38  
concentracion_vapor           5.37           5.42  
densidad_aire              1281.97        1282.08  
velocidad_viento              1.11           0.49  
direccion_viento             206.4          209.6  
precipitacion_total            0.0            0.0  
duracion_lluvia                  0              0

In [8]:
# Mostrar los últimos 5 registros
weather_df.tail()

medicion  presion_atmosferica  temperatura_celsius  \
52691  31/12/20 23:20               978.32                 2.28   
52692  31/12/20 23:30               978.30                 2.13   
52693  31/12/20 23:40               978.26                 1.99   
52694  31/12/20 23:50               978.26                 2.07   
52695   01/01/21 0:00               978.24                 2.01   

       temperatura_kelvin  humedad_relativa  presion_vapor  \
52691              277.16              80.0           5.76   
52692              277.01              83.1           5.92   
52693              276.88              82.2           5.80   
52694              276.95              81.4           5.77   
52695              276.89              82.4           5.82   

       humedad_especifica  concentracion_vapor  densidad_aire  \
52691                3.67                 5.89        1234.61   
52692                3.77                 6.05        1235.20   
52693                3.69                 5.93        1235.82   
52694                3.68                 5.90        1235.49   
52695                3.71                 5.95        1235.71   

       velocidad_viento  direccion_viento  precipitacion_total  \
52691              0.73             180.6                  0.0   
52692              0.43             174.0                  0.0   
52693              0.38             248.9                  0.0   
52694              0.57             196.6                  0.0   
52695              0.57             221.3                  0.0   

       duracion_lluvia  
52691                0  
52692                0  
52693                0  
52694                0  
52695                0

In [9]:
# Mostrar los ultimos 5 registros, pero transponiendo para poder observar todas las columnas del dataframe
weather_df.tail().T

52691           52692           52693  \
medicion             31/12/20 23:20  31/12/20 23:30  31/12/20 23:40   
presion_atmosferica          978.32           978.3          978.26   
temperatura_celsius            2.28            2.13            1.99   
temperatura_kelvin           277.16          277.01          276.88   
humedad_relativa               80.0            83.1            82.2   
presion_vapor                  5.76            5.92             5.8   
humedad_especifica             3.67            3.77            3.69   
concentracion_vapor            5.89            6.05            5.93   
densidad_aire               1234.61          1235.2         1235.82   
velocidad_viento               0.73            0.43            0.38   
direccion_viento              180.6           174.0           248.9   
precipitacion_total             0.0             0.0             0.0   
duracion_lluvia                   0               0               0   

                              52694          52695  
medicion             31/12/20 23:50  01/01/21 0:00  
presion_atmosferica          978.26         978.24  
temperatura_celsius            2.07           2.01  
temperatura_kelvin           276.95         276.89  
humedad_relativa               81.4           82.4  
presion_vapor                  5.77           5.82  
humedad_especifica             3.68           3.71  
concentracion_vapor             5.9           5.95  
densidad_aire               1235.49        1235.71  
velocidad_viento               0.57           0.57  
direccion_viento              196.6          221.3  
precipitacion_total             0.0            0.0  
duracion_lluvia                   0              0

**Valores Faltantes por columna**

Utilizando el método de Pandas isnull() combinado con el método de suma sum() podemos buscar cualquier valor faltante e identificar la columna en donde se encuentra.

In [10]:
# Verificar si hay valores faltantes en el dataframe por columna
print("Valores Faltantes por columna")
print("-" * 30)
print(weather_df.isnull().sum()) 

Valores Faltantes por columna
------------------------------
medicion               0
presion_atmosferica    0
temperatura_celsius    0
temperatura_kelvin     0
humedad_relativa       0
presion_vapor          0
humedad_especifica     0
concentracion_vapor    0
densidad_aire          0
velocidad_viento       0
direccion_viento       0
precipitacion_total    0
duracion_lluvia        0
dtype: int64


2. Determina el tipo de datos que tienen las columnas.
- Cambia la columna `medicion` a datetime.
- Obtén dos columnas adicionales separando `medicion` en `fecha` y `hora`.

In [11]:
# Revisar tipos de datos actuales
print("Tipos de datos antes de conversión:")
print(weather_df.dtypes)

Tipos de datos antes de conversión:
medicion                object
presion_atmosferica    float64
temperatura_celsius    float64
temperatura_kelvin     float64
humedad_relativa       float64
presion_vapor          float64
humedad_especifica     float64
concentracion_vapor    float64
densidad_aire          float64
velocidad_viento       float64
direccion_viento       float64
precipitacion_total    float64
duracion_lluvia          int64
dtype: object


In [ ]:
# No, the .to_datetime() method is a pandas function, not a DataFrame/Series method, and does not have an inplace argument.
# The correct way is:
weather_df['medicion'] = pd.to_datetime(weather_df['medicion'], format='%d/%m/%y %H:%M')

In [12]:
# Convertir la columna 'medicion' a tipo datetime
print(f"Tipo de dato previo a la conversion: {weather_df.medicion.dtypes}")
weather_df['medicion'] = pd.to_datetime(weather_df['medicion'], format='%d/%m/%y %H:%M')
print(f"Tipo de dato posterior a la conversion: {weather_df.medicion.dtypes}")

Tipo de dato previo a la conversion: object
Tipo de dato posterior a la conversion: datetime64[ns]


In [13]:
# Crear columnas adicionales 'fecha' y 'hora' a partir de 'medicion'
weather_df['fecha'] = weather_df['medicion'].dt.date
weather_df['hora'] = weather_df['medicion'].dt.time

In [19]:
# Verificar los cambios
print("Tipos de datos después de conversión:")
print("-" * 40)
print(weather_df.dtypes)


Tipos de datos después de conversión:
----------------------------------------
medicion               datetime64[ns]
presion_atmosferica           float64
temperatura_celsius           float64
temperatura_kelvin            float64
humedad_relativa              float64
presion_vapor                 float64
humedad_especifica            float64
concentracion_vapor           float64
densidad_aire                 float64
velocidad_viento              float64
direccion_viento              float64
precipitacion_total           float64
duracion_lluvia                 int64
fecha                          object
hora                           object
dtype: object


In [18]:
# Primeras filas con los cambios realizados y nuevas columnas
weather_df[['medicion', 'fecha', 'hora']].head()

medicion       fecha      hora
0 2020-01-01 00:10:00  2020-01-01  00:10:00
1 2020-01-01 00:20:00  2020-01-01  00:20:00
2 2020-01-01 00:30:00  2020-01-01  00:30:00
3 2020-01-01 00:40:00  2020-01-01  00:40:00
4 2020-01-01 00:50:00  2020-01-01  00:50:00

3. Determina la cantidad de valores únicos de cada columna.
- La columna `medicion` tiene un valor menos que el número total de filas del dataframe, lo que indica que hay una fecha duplicada. Identifica cuál es.
- Comprueba si las demás columnas también contienen los mismos valores; si es así, elimina  uno de los registros duplicados.


In [ ]:
weather_df = weather_df.drop_duplicates(subset=['medicion'], keep='first')



In [ ]:

# 3. Determina la cantidad de valores únicos de cada columna
print("Valores únicos por columna:")
print(weather_df.nunique())

# Identifica la fecha duplicada en 'medicion'
duplicado = weather_df['medicion'][weather_df['medicion'].duplicated()].iloc[0]
print(f"\nFecha duplicada en 'medicion': {duplicado}")

# Elimina el duplicado (sin checar si los valores son iguales)
weather_df = weather_df.drop_duplicates(subset=weather_df.columns.tolist(), keep='first')
print("Duplicado eliminado.")








In [ ]:
# Mostrar la cantidad de valores únicos por columna y comparar cada columna con la cantidad total de filas
print("Cantidad de valores únicos por columna:")
unicos = weather_df.nunique()
total_filas = len(weather_df)
for col, n_unicos in unicos.items():
    print(f"{col}: {total_filas - n_unicos} valores no únicos (total filas: {total_filas})")

Cantidad de valores únicos por columna:
medicion               52695
presion_atmosferica     5052
temperatura_celsius     3680
temperatura_kelvin      3836
humedad_relativa        5494
presion_vapor           2059
humedad_especifica      1337
concentracion_vapor     2087
densidad_aire          14826
velocidad_viento         977
direccion_viento        8099
precipitacion_total       42
duracion_lluvia           61
fecha                    367
hora                     144
dtype: int64


In [21]:
# Identificar el valor duplicado en 'medicion'
duplicados = weather_df['medicion'][weather_df['medicion'].duplicated(keep=False)]
duplicados

19043   2020-05-12 06:00:00
19044   2020-05-12 06:00:00
Name: medicion, dtype: datetime64[ns]

In [22]:
if not duplicados.empty:
    print("\nFechas/hora duplicadas en 'medicion':")
    print(duplicados)

    # Verificar si las demás columnas también son idénticas para esos duplicados
    duplicados_df = weather_df[weather_df['medicion'].isin(duplicados)]
    # Agrupar por 'medicion' y contar los duplicados exactos
    iguales = duplicados_df.groupby('medicion').apply(lambda x: x.drop('medicion', axis=1).nunique().max() == 1)
    print("\n¿Todos los valores de las otras columnas son idénticos para los duplicados?")
    print(iguales)

    # Si todos los valores son idénticos, eliminar duplicados conservando solo uno
    if iguales.all():
        weather_df = weather_df.drop_duplicates(subset=weather_df.columns.tolist(), keep='first')
        print("\nDuplicado eliminado (todas las columnas iguales).")
    else:
        print("\nDuplicados encontrados pero con diferencias en otras columnas. No se eliminan automáticamente.")
else:
    print("\nNo se encontraron duplicados en 'medicion'.")


Fechas/hora duplicadas en 'medicion':
19043   2020-05-12 06:00:00
19044   2020-05-12 06:00:00
Name: medicion, dtype: datetime64[ns]

¿Todos los valores de las otras columnas son idénticos para los duplicados?
medicion
2020-05-12 06:00:00    True
dtype: bool

Duplicado eliminado (todas las columnas iguales).


C:\Users\632464\AppData\Local\Temp\ipykernel_28656\890586825.py:8: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  iguales = duplicados_df.groupby('medicion').apply(lambda x: x.drop('medicion', axis=1).nunique().max() == 1)


4. Imprime nuevamente las cantidad de valores únicos por columna.
- ¿Por qué son 144 horas diferentes?
- La cantidad de combinaciones únicas de fecha y hora no coincide con el número total de mediciones. Aunque no haya valores NaN, esto no garantiza que no se hayan omitido registros. Identifica las fechas en las que faltan mediciones.
- Una vez concluido el análisis de mediciones faltantes, elimina todos los registros que no correspondan al año 2020 y la columna `medicion`.


In [ ]:
weather_df = weather_df[weather_df['medicion'].dt.year == 2020].drop(columns=['medicion'])
print(weather_df.head())
# Verificar que solo hay registros de 2020
print("Años únicos en el dataframe:", set(weather_df.index.year) if hasattr(weather_df.index, 'year') else "No hay columna de fechas en el índice")





5. Obtén un dataframe `indicadores_diarios` que sólo almacene el mayor valor de las mediciones por fecha para responder:
- ¿Cuál fue la máxima precipitación y a qué fecha pertenece?
- ¿Cuáles fueron los tres días con las velocidades de viento más altas registradas?
- ¿Cuántos días rebasaron el 90% de humedad relativa y los 30° celsius?

6. La función `describe()` resume múltiples estadísticas descriptivas. Utiliza esta función para explorar y responder las preguntas sobre las variables en el dataframe de `indicadores_diarios `:
- ¿Cuál es la desviación estándar de la concentración de vapor y qué indica?
- ¿Cuál es la mediana de la presión atmosférica y qué significa?
- ¿Cuál es el tercer cuartil de la duración de la lluvia y cómo se interpreta?

7.	Dibuja un histograma para cada una de las columnas de `indicadores_diarios`.

8. Utiliza el dataframe de indicadores diarios para crear un nuevo dataframe `indicadores_mensuales`. Para ello:
- Crea una nueva columna `mes` a partir de la fecha que se encuentra en el índice.
- Haciendo uso de `groupby()`, calcula los promedios mensuales y conserva únicamente las columnas `humedad_relativa`, `temperatura_celsius` y `velocidad_viento` en el nuevo dataframe.

9. Construye el mismo dataframe mensual, pero esta vez utilizando la función `pivot_table()`.
- Añade la columna `indice_calor` usando la fórmula de Thom (*Temperature - Humidity Index*, TDI):
> `temperatura_celsius + 0.33 x humedad_relativa - 0.7 x velocidad_viento - 4`

10. Recuerda que de las cuatro funciones estudiadas para manipular la estructura del dataframe:
- `melt/pivot` hacen las transformaciones de manera controlada, es decir puedes
definir por medio de sus parámetros qué variables quedarán como índices, columnas y valores.
- `stack/unstack` la conversión se aplica siempre sobre los niveles inferiores de index/columns.
- Convierte el dataframe del ejercicio anterior a formato largo usando: `melt()` y `stack()` y comenta las diferencias.